src/poetry/puzzle/provider.py

```diff
@@ -268,18 +269,33 @@ def search_for(self, dependency: Dependency) -> list[DependencyPackage]:
             return PackageCollection(dependency, [self._package])
 
         if dependency.is_direct_origin():
-            packages = [self.search_for_direct_origin_dependency(dependency)]
+            package = self.search_for_direct_origin_dependency(dependency)
+            self._direct_origin_packages[dependency.name] = package
+            return PackageCollection(dependency, [package])
 
-        else:
-            packages = self._pool.find_packages(dependency)
-
-            packages.sort(
-                key=lambda p: (
-                    not p.is_prerelease() and not dependency.allows_prereleases(),
-                    p.version,
-                ),
-                reverse=True,
+        # If we've previously found a direct-origin package that meets this dependency,
+        # use it.
+        #
+        # We rely on the VersionSolver resolving direct-origin dependencies first.
+        direct_origin_package = self._direct_origin_packages.get(dependency.name)
+        if direct_origin_package is not None:
+            package = direct_origin_package.with_features(dependency.features)
+            packages = (
+                [package]
+                if package.satisfies(dependency, ignore_source_type=True)
+                else []
             )
```

Comment:
```suggestion
            packages = (
                [direct_origin_package]
                if dependency.constraint.allows(direct_origin_package.version)
                else []
            )
```

If I am right, `search_for` did always return packages without features/extras so far and `complete_package` added these. Maybe, we should keep this pattern?

Reply:
I'm ok with that